natlib/svcd.c

```diff
@@ -185,3 +196,208 @@ static int svcd_init( lua_State *L )
 
     return 0;
 }
+
+//////////////////////////////////////////////////////////////////////////////
+// SVCD.write((string targetip, number svcid, number attrid, lightfunction payload, 
+//                        number timeout_ms, lightfunction on_done)) 
+// Authors: Aparna Dhinakaran, David Ho, Romi Phadte, Cesar Torres
+/////////////////////////////////////////////////////////////
+
+
+void resolve_table(lua_State *L, char* key){
+    // push the global table, resolve it and then pop it off the stack
+    lua_getglobal(L, "SVCD");
+    lua_pushstring(L, key);
+    lua_gettable(L, -2);
+    lua_insert(L, -2);
+    lua_settop(L, -2);
+}
+
+static int svcd_write_invoke_later(lua_State *L);
+static int svcd_callback_handler(lua_State *L);
+static int anon_function(lua_State *L);
+
+static int svcd_callback_handler(lua_State *L){
+    int ivkd = (int) lua_tonumber(L, lua_upvalueindex(1));
+    printf("making callback function using ivkd = %i \n", ivkd);
+
+    resolve_table(L, "TIMEOUT");
+    int timeout = (int) lua_tonumber(L, -1);
+
+    resolve_table(L, "handlers"); // 7
+    lua_pushnumber(L, ivkd);
+    lua_gettable(L, -2);
+
+    int isnil = (int) lua_toboolean(L, -1);
+    if(isnil != 0){
+        lua_pushnumber(L, timeout);
+        lua_call(L, 1, 0);
+
+        resolve_table(L, "handlers"); // 7
+        lua_pushnumber(L, ivkd);
+        lua_pushnil(L);
+        lua_settable(L, -3);
+    }
+    return 0;
+}
+static int anon_function(lua_State *L){
+    int ivkd = (int)lua_tonumber(L, 1);
+    printf("making anon function using ivkd = %i \n", ivkd);
+
+    lua_pushnumber(L, ivkd);
+    lua_pushcclosure(L, &svcd_callback_handler, 0);
+    return 1;
+}
+
+// write 
+static int svcd_write_invoke_later(lua_State *L){
+    int ivkd = (int)lua_tonumber(L, 1);
+    int timeout_ms = (int)lua_tonumber(L, 2);
+    lua_pushlightfunction(L, libstorm_os_invoke_later);
+    lua_pushnumber(L, timeout_ms * MILLISECOND_TICKS);
+    
+    printf("making custom anon funtion\n");
+    // lua_pushlightfunction(L, anon_function);
+    lua_getglobal(L, "anon_function");
+        lua_pushnumber(L, ivkd);
+        lua_call(L, 0, 1);
+    printf("made custom anon funtion\n");
+
+    lua_call(L, 2, 0);
+    return 0;
+}
+
+static int svcd_write( lua_State *L )
+{   
+    // REGISTER CLUA FUNCTIONS IN GLOBAL SPACE
+    // lua_pushcfunction(L, svcd_write_invoke_later);
+    // lua_setglobal(L, "svcd_write_invoke_later");
+    // lua_pushcfunction(L, anon_function);
+    // lua_setglobal(L, "anon_function");
+    // lua_pushcfunction(L, svcd_callback_handler);
+    // lua_setglobal(L, "svcd_callback_handler");
+    // END REGISTER
+
+    printf("Sexy write!\n");    
```

Comment:
I would like to echo the sentiments of Holy Cow. All these print statements are making your code hard to read efficiently. Feel free to use them for testing your code, but remove them when you push it here. 

Best, 

- PB for your Jelly or, The Perfect Storm